File: desktop-background/interfaces/desktop-background.py
UTILITY_EXECUTABLE_PATH = "../desktop-background.exe"

import subprocess
import asyncio
import os
import json

async def example():
  print("desktop_background example - Python")

  update_utility_executable_path("../desktop-background.exe")

  p = await get_current_desktop_background_image()
  print("Current desktop background image:")
  print(p)
  p = await set_current_desktop_background_image(p)
  print("Updated the desktop background image:")
  print(p)
  

def update_utility_executable_path(exeFilePath):
  """
  Update the path for the utility executable.
  """
  global UTILITY_EXECUTABLE_PATH
  UTILITY_EXECUTABLE_PATH = exeFilePath


async def get_current_desktop_background_image():
  """
  Retrieves the state of the foreground window.
  """
  text = await execute_window_state_utility(["--get"])
  if len(text) <= 2 or not text.startswith("{") or not text.endswith("}"):
    raise ValueError(f"Desktop background utility failed unexpectedly: {json.dumps(text)}")
  obj = json.loads(text)
  if not obj or not "path" in obj:
    raise ValueError(f"Desktop background utility failed unexpectedly: {json.dumps(text)}")
  return obj.path

async def set_current_desktop_background_image(target):
  """
  Retrieves the state of the foreground window.
  """
  text = await execute_window_state_utility(["--set", target])
  if len(text) <= 2 or not text.startswith("{") or not text.endswith("}"):
    raise ValueError(f"Desktop background utility failed unexpectedly: {json.dumps(text)}")
  obj = json.loads(text)
  if not obj or not "path" in obj:
    raise ValueError(f"Desktop background utility failed unexpectedly: {json.dumps(text)}")
  return obj.path


async def execute_window_state_utility(args):
  if not isinstance(UTILITY_EXECUTABLE_PATH, str):
    raise TypeError("Executable file path must be a string")

  if not os.path.exists(UTILITY_EXECUTABLE_PATH):
    raise FileNotFoundError(
      f"Executable file '{UTILITY_EXECUTABLE_PATH}' not found."
    )
    
  # Convert number and boolean arguments to strings
  args = [str(arg) if not isinstance(arg, (str, bytes)) else arg for arg in args]

  if not all(isinstance(arg, str) for arg in args):
    raise TypeError("Arguments must be a list of strings")


  command = UTILITY_EXECUTABLE_PATH
  process = await asyncio.create_subprocess_exec(
    command, *args, stdout=subprocess.PIPE, stderr=subprocess.PIPE
  )

  stdout, stderr = await process.communicate()

  if process.returncode == 0:
    return stdout.decode("utf-8")

  raise Exception(
    stderr.decode("utf-8").strip() or f"Error code {process.returncode}"
  )

# asyncio.run(example())

File: desktop-background/interfaces/desktop-background.ts
let UTILITY_EXECUTABLE_PATH = "../desktop-background.exe";

import child_process from "node:child_process";
// import {Buffer} from "node:buffer";

async function example() {
  console.log("desktop-background example - Typescript");

  updateUtilityExecutablePath("../desktop-background.exe");

  let p = await getCurrentDesktopBackgroundImage();
  console.log("Current desktop background image:");
  console.log(p);
  p = await setCurrentDesktopBackgroundImage(p);
  console.log("Updated the desktop background image:");
  console.log(p);
}

/**
 * Update the path for the utility executable.
 */
function updateUtilityExecutablePath(exeFilePath: string) {
  UTILITY_EXECUTABLE_PATH = exeFilePath;
}

/**
 * Get the current desktop background image file path
 */
export async function getCurrentDesktopBackgroundImage(): Promise<string> {
  const text = await executeWindowStateUtility(["--get"]);
  if (text.length <= 2 || !text.startsWith("{") || !text.endsWith("}")) {
    throw new Error(
      `Desktop background utility failed: ${JSON.stringify(text)}`
    );
  }
  const obj: { error: string } | { path: string } = JSON.parse(text);
  if ("error" in obj) {
    throw new Error(
      `Desktop background utility returned an error: ${obj.error}`
    );
  }
  if ("path" in obj) {
    return obj.path;
  }
  throw new Error(`Desktop background utility failed unexpectedly: ${text}`);
}

/**
 * Retrieves the states of children windows from the desktop.
 * @returns An array of window data for each child window.
 */
export async function setCurrentDesktopBackgroundImage(
  target: string
): Promise<string> {
  const text = await executeWindowStateUtility(["--set", target]);
  if (text.length <= 2 || !text.startsWith("{") || !text.endsWith("}")) {
    throw new Error(
      `Desktop background utility failed: ${JSON.stringify(text)}`
    );
  }
  const obj: { error: string } | { path: string } = JSON.parse(text);
  if ("error" in obj) {
    throw new Error(
      `Desktop background utility returned an error: ${obj.error}`
    );
  }
  if ("path" in obj) {
    return obj.path;
  }
  throw new Error(`Desktop background utility failed unexpectedly: ${text}`);
}

/**
 * Execute the utility process with specified arguments
 */
function executeWindowStateUtility(args: string[]): Promise<string> {
  return new Promise((resolve, reject) => {
    try {
      const command = UTILITY_EXECUTABLE_PATH;
      const child = child_process.spawn(command, args, {
        shell: false,
        stdio: ["ignore", "pipe", "pipe"],
      });
      const chunks: Buffer[] = [];
      child.stdout.on("data", (data) => chunks.push(data));
      child.stderr.on("data", (data) => chunks.push(data));
      child.on("error", (err: any) =>
        reject(
          err.code === "ENOENT"
            ? new Error(`Could not find executable at "${command}"`)
            : err
        )
      );
      child.on("exit", (exit) => {
        const text = Buffer.concat(chunks).toString("utf8").trim();
        return exit === 0
          ? resolve(text)
          : reject(new Error(text || `Error code ${exit}`));
      });
    } catch (err) {
      reject(err);
    }
  });
}

// example();
